### PR TITLE
IOTE-4566 Limit temperature setpoint ranges

### PR DESCRIPTION
**Pull request by nickolas-deboom**
# Type of Change

- [ ] WWST Certification Request
     - If this is your first time contributing code:
          - [ ] I have reviewed the README.md file
          - [ ] I have reviewed the CODE_OF_CONDUCT.md file
          - [ ] I have signed the CLA
     - [ ] I plan on entering a WWST Certification Request or have entered a request through the WWST Certification console at developer.smartthings.com
- [x] Bug fix
- [ ] New feature
- [ ] Refactor

# Checklist

- [x] I have performed a self-review of my code
- [x] I have commented my code in hard-to-understand areas
- [x] I have verified my changes by testing with a device or have communicated a plan for testing
- [x] I am adding new behavior, such as adding a sub-driver, and have added and run new unit tests to cover the new behavior

# Description of Change

[IOTE-4566](https://smartthings.atlassian.net/browse/IOTE-4566)

Clamp the min and max values of temperature setpoint capabilities so that the ranges of deg C and deg F don't overlap. This is currently being done for the oven device type and now should be extended to other devices types using the `temperatureSetpoint`, `thermostatHeatingSetpoint`, or `thermostatCoolingSetpoint` capabilities (or the associated Setpoint Range capabilities). These changes are necessary because the driver is not aware of the temperature unit when a setpoint value is updated via a capability command. Threshold values were chosen depending on the device type as follows:

Dishwasher:
- Threshold: 90 degrees
- Range: 33C - 90C / 91.4F - 194F

Laundry Dryer:
- Threshold: 80 degrees
- Range: 27C - 80C / 80.6F - 176F

Laundry Washer:
- Threshold: 55 degrees
- Range: 13C - 55C / 55.4F - 131F

Refrigerator:
- Threshold: 20 degrees
- Range: -6C - 20C / 21.2F - 68F

Thermostat / Room AC:
- Threshold: 40 degrees
- Range: 5C - 40C / 41F - 104F

Please let me know if you have any opinions on better values to use.

# Summary of Completed Tests

See new unit tests.

Testing with VDA:
- For each device, verify the correct range is present in both deg C and F, and that attempting to exceed the bounds in either direction results in the app displaying "Out of range". Note that the range that displays in the ST app will not necessarily be the max range defined in the drivers.
- [x] Dishwasher *
    - ~~On VDA v1.3.4, this device crashes a few seconds after onboarding and the device appears as offline in the ST app.~~
        - This is fixed in v1.3.5.
    - The plugin only allows the setpoint value to go up to 100 *F, even though the range displays correctly as 91.4 *F - 194.0 *F. The '+' becomes grayed out and the setpoint cannot be increased further. The slider works correctly in *C. This issue is tracked by MTR-842.
- [x] Laundry Dryer *
- [x] Laundry Washer *
- [x] Refrigerator *
    - ~~On VDA v1.3.4, this device crashes a few seconds after onboarding and the device appears as offline in the ST app.~~
    - This is fixed in v1.3.5.
- [x] Thermostat / Room AC
    - VDA Thermostat device sends Null values for the MinMeasuredValue and MaxMeasuredValue attributes of the temperatureMeasurement cluster. The temperature range displays as the plugin default of -20 to 50 *C, but this functionality is covered by unit tests.

\* For these device types, the plugin is not correctly limiting the setpoint sliders from exceeding the allotted range. The minimum and maximum allowed values appear as expected, but the user is able to go beyond the range using the '-' and '+' buttons, which are not grayed out when the slider is at the min or max as they should be. Also, the slider can sometimes be moved beyond the range with sliding as well. The thermostat device type does not have this issue.  This issue is tracked by MTR-841.

[IOTE-4566]: https://smartthings.atlassian.net/browse/IOTE-4566?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua [hunk 1]
```diff
@@ -15,29 +15,36 @@
...
 
-local REFRIGERATOR_DEVICE_TYPE_ID = 0x0070
-local TEMPERATURE_CONTROLLED_CABINET_DEVICE_TYPE_ID = 0x0071
+local COMPONENT_TO_ENDPOINT_MAP = "__component_to_endpoint_map"
+local SUPPORTED_TEMPERATURE_LEVELS_MAP = "__supported_temperature_levels_map"
+local SUPPORTED_REFRIGERATOR_TCC_MODES_MAP = "__supported_refrigerator_tcc_modes_map"
+
+-- This is a work around to handle when units for temperatureSetpoint is changed for the App.
+-- When units are switched, we will never know the units of the received command value as the arguments don't contain the unit.
+-- So to handle this we assume the following ranges considering usual laundry temperatures:
```
Comment (ctowns) [inline drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:37]:
```suggestion
-- So to handle this we assume the following ranges considering usual refrigerator temperatures:
```

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua [hunk 2]
```diff
@@ -56,6 +63,18 @@ local function component_to_endpoint(device, component)
   return 1
 end
 
+local function is_matter_refrigerator(opts, driver, device)
```
Comment (ctowns) [inline drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:66]: nit: why is this moved?

File: drivers/SmartThings/matter-appliance/src/matter-laundry/init.lua [hunk 3]
```diff
@@ -100,6 +143,84 @@ local function supported_temperature_levels_attr_handler(driver, device, ib, res
...
+    -- We clamp the max and min values as per the assumed oven temperature ranges.
+    if val < min_temp_in_c or val > max_temp_in_c then
+      if limit_field == setpoint_limit_device_field.MIN_TEMP then
+        val = min_temp_in_c
+      else
+        val = max_temp_in_c
+      end
+    end
+
+    log.info("Setting " .. field .. " to " .. string.format("%s", val))
```
Comment (ctowns) [inline drivers/SmartThings/matter-appliance/src/matter-laundry/init.lua:219]: I would use device.log so that the device info is automatically prepended to the log
```suggestion
    device.log.info("Setting " .. field .. " to " .. string.format("%s", val))
```

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua [hunk 4]
```diff
@@ -15,29 +15,36 @@
...
-local REFRIGERATOR_DEVICE_TYPE_ID = 0x0070
-local TEMPERATURE_CONTROLLED_CABINET_DEVICE_TYPE_ID = 0x0071
+local COMPONENT_TO_ENDPOINT_MAP = "__component_to_endpoint_map"
+local SUPPORTED_TEMPERATURE_LEVELS_MAP = "__supported_temperature_levels_map"
+local SUPPORTED_REFRIGERATOR_TCC_MODES_MAP = "__supported_refrigerator_tcc_modes_map"
+
+-- This is a work around to handle when units for temperatureSetpoint is changed for the App.
+-- When units are switched, we will never know the units of the received command value as the arguments don't contain the unit.
+-- So to handle this we assume the following ranges considering usual laundry temperatures:
+--   1. if the received setpoint command value is in range -6 ~ 20, it is inferred as *C
```
Comment (ctowns) [inline drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:38]: I think this range would work well for refrigerator, but I think we'd need to be able to have the range go down to 0 degF/-18 degC (give or take a few degrees) for freezers. 0 degF is the USDA recommended temp for a freezer: https://ask.usda.gov/s/article/What-is-the-right-temperature-for-the-freezer. 

Using the component to endpoint map and vice versa, you can determine if the setpoint value is received from the freezer component and then adjust the values from there.

**Comment by LEE-HYOJUNG27** [comment]
I think you need to modify the temperature range for each device type. Because Celsius and Fahrenheit written in the description do not actually match and can cause conflict.

**[Example of conflict]**

Laundry Dryer:
Threshold: 80 degrees
Range: 26C - 80C / 81F - 178F

**>> Range: 26C - 80C / 78.8F - 176F 
>> Conflict range : 78.8F - 80F** 

**[Recalculated range]**

Dishwasher:
Threshold: 90 degrees
Range: 23C - 90C / 91F - 194F

Laundry Dryer:
Threshold: 80 degrees
Range: 27C - 80C / 80.6F - 176F

Laundry Washer:
Threshold: 55 degrees
Range: 13C - 55C / 55.4F - 131F

Refrigerator:
Threshold: 20 degrees
Range: -6C - 20C / 21.2F - 68F

Thermostat / Room AC:
Threshold: 39 degrees
Range: 5C - 39C / 41F - 102.2F

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua (hunk 4 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:38]: I originally implemented in this way, but then looked at the `refrigerator-freezer-tn-tl` profile and saw that the freezer did not include the `temperatureSetpoint` capability. I assumed this was the case for all the profiles so I only implemented the temperature range limits for the refrigerator. But, it looks like `temperatureSetpoint` does exist in 2/4 of the refrigerator profiles, so I will make the change you suggested. It's clear why it is not in `refrigerator-freezer-tl.yml`, but any idea why it is not in `refrigerator-freezer-tn-tl.yml`?

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua (hunk 2 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:66]: I thought it would make sense to move it above the 'Matter Handlers' section, since it doesn't really fit there!

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua (hunk 1 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua]: Fixed!

File: drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua (hunk 4 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-appliance/src/matter-refrigerator/init.lua:39]: Implemented this in b6c2b0853c580cc6ae4bedde861a8b78d7b8e274!

**Comment by nickolas-deboom** [comment]
> I think you need to modify the temperature range for each device types. Because Celsius and Fahrenheit written in the description do not actually match and can cause conflict.
> 
> **[Example of conflict]**
> 
> Laundry Dryer: Threshold: 80 degrees Range: 26C - 80C / 81F - 178F
> 
> **-> Range: 26C - 80C / 78.8F - 176F -> Conflict range : 78.8F - 80F**
> 
> **[Recalculated range]**
> 
> Dishwasher: Threshold: 90 degrees Range: 33C - 90C / 91.4F - 194F
> 
> Laundry Dryer: Threshold: 80 degrees Range: 27C - 80C / 80.6F - 176F
> 
> Laundry Washer: Threshold: 55 degrees Range: 13C - 55C / 55.4F - 131F
> 
> Refrigerator: Threshold: 20 degrees Range: -6C - 20C / 21.2F - 68F
> 
> Thermostat / Room AC: Threshold: 39 degrees Range: 5C - 39C / 41F - 102.2F
> 
> cc: @HunsupJung

I applied these new ranges in b6c2b0853c580cc6ae4bedde861a8b78d7b8e274!

**Review by HunsupJung — APPROVED** [review]

File: drivers/SmartThings/matter-appliance/src/matter-dishwasher/init.lua [hunk 5]
```diff
@@ -88,6 +126,62 @@ local function supported_temperature_levels_attr_handler(driver, device, ib, res
...
+      { feature_bitmap = clusters.TemperatureControl.types.Feature.TEMPERATURE_NUMBER })
+    if #tn_eps == 0 then
+      device.log.warn(string.format("Device does not support TEMPERATURE_NUMBER feature"))
+      return
+    end
+    local field = string.format("%s-%d", limit_field, ib.endpoint_id)
+    local val = ib.data.value / 100.0
+
+    -- We clamp the max and min values as per the assumed oven temperature ranges.
+    if val < DISHWASHER_MIN_TEMP_IN_C or val > DISHWASHER_MAX_TEMP_IN_C then
```
Comment (ctowns) [inline drivers/SmartThings/matter-appliance/src/matter-dishwasher/init.lua:172]: I think you could simplify this by using `utils.clamp_value()` here and in the other places where this logic is used. Example from matter-switch [here](https://github.com/SmartThingsCommunity/SmartThingsEdgeDrivers/blob/cf148fe773ea89daa85584e33ebffcf1dbd5e543/drivers/SmartThings/matter-switch/src/init.lua#L311).

**Review by ctowns — APPROVED** [review]
Left a small comment for a potential code improvement, but this looks good to me!

**Review by hcarter-775 — APPROVED** [review]

File: drivers/SmartThings/matter-appliance/src/matter-dishwasher/init.lua (hunk 5 above)
Comment (nickolas-deboom) [reply drivers/SmartThings/matter-appliance/src/matter-dishwasher/init.lua]: Great suggestion! I updated it in each driver that used this logic in the latest commit
